packages/web-extension/src/walletManager/walletManagerWorker.ts

```diff
@@ -11,48 +21,85 @@ import {
   WalletManagerApi,
   WalletManagerProps
 } from './walletManager.types';
-import { consumeKeyAgent } from '../keyAgent';
+import { WalletRepository } from './WalletRepository';
+import { consumeSignerManagerApi } from './SignerManager/consumeSignerManagerApi';
 import { observableWalletProperties } from '../observableWallet';
 import { walletChannel } from './util';
 
-export interface WalletManagerDependencies {
+const getWalletStoreId = (walletId: WalletId, chainId: Cardano.ChainId, accountIndex?: number): string => {
+  // Index 0 should be backwards compatible (so it should never have the account index concatenated).
+  if (accountIndex !== undefined && accountIndex !== 0) {
+    return `${chainId.networkId}-${chainId.networkMagic}-${walletId}-${accountIndex}`;
+  }
+
+  return `${chainId.networkId}-${chainId.networkMagic}-${walletId}`;
+};
+
+export interface WalletManagerDependencies<Metadata extends {}> {
   walletFactory: WalletFactory;
   storesFactory: StoresFactory;
   managerStorage: Storage.StorageArea;
+  walletRepository: WalletRepository<Metadata>;
 }
 
 /**
  * Helper class for background scripts using wallet manager.
  * Uses wallet and store factories to create wallets.
  * Keeps track of created stores and reuses them when a wallet is reactivated.
  */
-export class WalletManagerWorker implements WalletManagerApi {
+export class WalletManagerWorker<Metadata extends {}> implements WalletManagerApi {
+  chainId$: Observable<Cardano.ChainId>;
   activeWallet$: Observable<ObservableWallet>;
 
-  #activeWalletId: string | null = null;
+  #activeWalletProps: WalletManagerActivateProps | null = null;
   #api$ = new BehaviorSubject<ObservableWallet | null>(null);
   #hostSubscription: Shutdown;
-  #keyAgentSubscription: Shutdown;
   #walletStores = new Map<string, storage.WalletStores>();
 
   #walletFactory: WalletFactory;
   #storesFactory: StoresFactory;
+  #walletRepository: WalletRepository<Metadata>;
+  #signerManagerApi: SignerManagerSignApi<Metadata> & Shutdown;
+
   #logger: Logger;
   #runtime: MinimalRuntime;
   #managerStorageKey: string;
   #managerStorage: Storage.StorageArea;
 
   constructor(
     { walletName }: WalletManagerProps,
-    { walletFactory, storesFactory, logger, runtime, managerStorage }: MessengerDependencies & WalletManagerDependencies
+    {
+      walletFactory,
+      storesFactory,
+      logger,
+      runtime,
+      managerStorage,
+      walletRepository
+    }: MessengerDependencies & WalletManagerDependencies<Metadata>
   ) {
+    this.#walletRepository = walletRepository;
     this.activeWallet$ = this.#api$.pipe(filter(isNotNil));
+    this.chainId$ = this.activeWallet$.pipe(
+      switchMap((wallet) =>
+        wallet.genesisParameters$.pipe(
+          map(
+            (params): Cardano.ChainId => ({
+              networkId: params.networkId,
+              networkMagic: params.networkMagic
+            })
+          )
+        )
+      )
```

Comment:
It would be a better/more optimal behavior, because currently it will re-emit same chain id when you activate different wallet on same network
```suggestion
      ),
      distinctUntilChanged(deepEquals)
```

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9